### PR TITLE
[TLI] Use the VFABI demangling when declaring vector variants.

### PR DESCRIPTION
**Pull request by labrinea**
When creating a declaration for a vector variant, in order to determine the argument types we need to consult the VFABI demangler. This will allow us to add TLI mappings with linear arguments (see #76060).

### CODE REVIEW

**Comment by labrinea** [comment]
Unfortunately the test cannot be autogenerated since the script can create check lines for globals and attributes but not for declarations.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 1]
```diff
@@ -37,20 +37,39 @@ STATISTIC(NumCompUsedAdded,
 /// vectorizes the CallInst CI with a vectorization factor of VF
 /// lanes. The TLI assumes that all parameters and the return type of
 /// CI (other than void) need to be widened to a VectorType of VF
-/// lanes.
+/// lanes unless the VFABI says otherwise.
```
Comment (mgabka) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:40]: Could you rephrase this sentence ?
I think like something like:
"For each mapping TLI contains a VABI prefix, which contains all information required to create vector function declaration."

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 2]
```diff
@@ -37,20 +37,39 @@ STATISTIC(NumCompUsedAdded,
...
 static void addVariantDeclaration(CallInst &CI, const ElementCount &VF,
-                                  bool Predicate, const StringRef VFName) {
+                                  const VecDesc *VD) {
   Module *M = CI.getModule();
 
   // Add function declaration.
   Type *RetTy = ToVectorTy(CI.getType(), VF);
```
Comment (mgabka) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:46]: is this going to work if function returns void?

it may also be worth to move it below the call to "tryDemangleForVFABI" which will ensure that all the VFABI rules are applied

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 3]
```diff
@@ -37,20 +37,39 @@ STATISTIC(NumCompUsedAdded,
...
   SmallVector<Type *, 4> Tys;
-  for (Value *ArgOperand : CI.args())
-    Tys.push_back(ToVectorTy(ArgOperand->getType(), VF));
+
+  const std::optional<VFInfo> Info = VFABI::tryDemangleForVFABI(
+      VD->getVectorFunctionABIVariantString(), CI.getFunctionType());
+
+  assert(Info && Info->Shape.VF == VF && "Mangled name does not match VF");
```
Comment (mgabka) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:52]: I think could be worth to split it into multiple asserts, so it is more clear if the issue is with tryDemangleForVFABI returning no value, i,e mangled prefix in the TLI is incorrect, or the VF does not match.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 4]
```diff
@@ -37,20 +37,39 @@ STATISTIC(NumCompUsedAdded,
...
+
+  for (VFParameter Param : Info->Shape.Parameters) {
```
Comment (mgabka) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:54]: looks like you could just use the "VFABI:CreateFunctionType" as it does what you need.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 1 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 2 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: It does work, but I've replaced the whole block of code with `VFABI:createFunctionType` as suggested.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 3 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 4 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done

**Review by mgabka — APPROVED** [review]

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 5]
```diff
@@ -35,35 +35,36 @@ STATISTIC(NumCompUsedAdded,
...
-/// lanes. The TLI assumes that all parameters and the return type of
-/// CI (other than void) need to be widened to a VectorType of VF
-/// lanes.
+/// lanes. For each mapping TLI contains a VABI prefix, which contains
+/// all information required to create vector function declaration.
 static void addVariantDeclaration(CallInst &CI, const ElementCount &VF,
-                                  bool Predicate, const StringRef VFName) {
+                                  const VecDesc *VD) {
   Module *M = CI.getModule();
+  FunctionType *FTy = CI.getFunctionType();
```
Comment (paschalis-mpeis) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:43]: nit: I would rename this to `ScalarFTy` to match similar code that uses VFABI.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 6]
```diff
@@ -35,35 +35,36 @@ STATISTIC(NumCompUsedAdded,
...
+
+  const std::optional<VFInfo> Info =
+      VFABI::tryDemangleForVFABI(VD->getVectorFunctionABIVariantString(), FTy);
+
+  assert(Info && "Failed to demangle vector variant");
+
+  assert(Info->Shape.VF == VF && "Mangled name does not match VF");
+
+  const StringRef VFName = VD->getVectorFnName();
+  FunctionType *VFTy = VFABI::createFunctionType(*Info, FTy);
```
Comment (paschalis-mpeis) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:55]: nit: I would rename this to `VectorFTy` to match similar code that uses VFABI.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 7]
```diff
@@ -35,35 +35,36 @@ STATISTIC(NumCompUsedAdded,
...
-  Function *VectorF =
-      Function::Create(FTy, Function::ExternalLinkage, VFName, M);
-  VectorF->copyAttributesFrom(CI.getCalledFunction());
+  assert(!FTy->isVarArg() && "VarArg functions are not supported.");
+
+  const std::optional<VFInfo> Info =
+      VFABI::tryDemangleForVFABI(VD->getVectorFunctionABIVariantString(), FTy);
+
+  assert(Info && "Failed to demangle vector variant");
+
```
Comment (paschalis-mpeis) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:51]: nit: could remove that new line.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 8]
```diff
@@ -35,35 +35,36 @@ STATISTIC(NumCompUsedAdded,
 
 /// A helper function that adds the vector function declaration that
 /// vectorizes the CallInst CI with a vectorization factor of VF
```
Comment (paschalis-mpeis) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:37]: nit: could improve original doxygen comment with `\p CI` instead of `CI`.
Same for `VF`.

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp [hunk 9]
```diff
@@ -35,35 +35,36 @@ STATISTIC(NumCompUsedAdded,
...
-/// lanes. The TLI assumes that all parameters and the return type of
-/// CI (other than void) need to be widened to a VectorType of VF
-/// lanes.
+/// lanes. For each mapping TLI contains a VABI prefix, which contains
```
Comment (paschalis-mpeis) [inline llvm/lib/Transforms/Utils/InjectTLIMappings.cpp:38]: nit: I would rewrite maybe to something like:
For each mapping, TLI provides/has a VABI prefix, which contains ...

**Review by paschalis-mpeis — APPROVED** [review]

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 5 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 6 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 7 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 8 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done

File: llvm/lib/Transforms/Utils/InjectTLIMappings.cpp (hunk 9 above)
Comment (labrinea) [reply llvm/lib/Transforms/Utils/InjectTLIMappings.cpp]: Done
